src/compiler/crystal/zero_one_or_many.cr

```diff
@@ -0,0 +1,110 @@
+# An Array(T)-like type that's optimized for the case of
+# frequently having zero or one elements.
+#
+# It's a struct, so if you have a ZeroOneOrMany and you
+# want to mutate, you have to use `with` or `without`
+# and overwrite the value you have with what you get
+# from those methods.
+#
+# Warning: you should never keep the original value
+# around when using the above methods as the internal
+# array might be mutated (for performance reasons),
+# so the original value might mutate in that case!
+#
+# TODO: add tests for this type.
+struct Crystal::ZeroOneOrMany(T)
+  include Enumerable(T)
+
+  getter value : Nil | T | Array(T)
+
+  def initialize
+    @value = nil
+  end
+
+  def initialize(@value : T)
+  end
+
+  def initialize(values : Array(T))
+    @value =
+      case values.size
+      when 0
+        nil
+      when 1
+        values.first
+      else
+        values
+      end
+  end
+
+  def each(& : T ->)
+    value = @value
+    case value
+    in Nil
+      # Nothing to do
+    in T
+      yield value
+    in Array(T)
+      value.each do |element|
+        yield element
+      end
+    end
+  end
+
+  def size : Int32
+    value = @value
+    case value
+    in Nil
+      0
+    in T
+      1
+    in Array(T)
+      value.size
+    end
+  end
+
+  def with(element : T) : ZeroOneOrMany(T)
+    value = @value
+    case value
+    in Nil
+      ZeroOneOrMany(T).new(element)
+    in T
+      ZeroOneOrMany(T).new([value, element] of T)
+    in Array(T)
+      value.push element
+      self
+    end
+  end
+
+  def with(elements : Enumerable(T)) : ZeroOneOrMany(T)
+    value = @value
+    case value
+    in Nil
+      ZeroOneOrMany(T).new(elements.map(&.as(T)))
```

Comment:
That's a good point. I think for that we need to have `elements` be an `Indexable`, and for that we need `ZeroOneOrMany` to be `Indexable`. But that's totally doable.

Reply:
Done!

Reply:
This was a great catch! This `with(elements)` is called when a `Call` is bound to its methods. The methods list is an Array, so this ended up creating an array inside `ZeroOneOrMany`. Now that we check `elements.size` we keep a single element here.

Using the benchmark I've been using in this PR:

Before: 3.25  (307.39ms) (±15.88%)  252MB/op
After: 3.39  (295.20ms) (±13.66%)  249MB/op

So that's like 1% less memory. In concrete numbers it's 3MB less memory here, which is a lot of memory for such a simple change!